jetty-documentation/src/main/java/org/eclipse/jetty/docs/JettyIncludeExtension.java

```diff
@@ -0,0 +1,180 @@
+ *   <dt>args</dt>
+ *   <dd>Optional, specifies the arguments to use in a Jetty server run.
+ *   The output produced by this run is included in the Asciidoc document.</dd>
+ *   <dt>highlight</dt>
+ *   <dd>Optional, specifies a regular expression that matches lines that should be highlighted.
+ *   If the regular expression contains capturing groups, only the text matching
+ *   the groups is highlighted, not the whole line.
+ *   </dd>
+ * </dl>
+ *
+ * @see JettyHomeTester
+ */
+public class JettyIncludeExtension implements ExtensionRegistry
+{
+    public void register(Asciidoctor asciidoctor)
+    {
+        asciidoctor.javaExtensionRegistry().includeProcessor(JettyIncludeProcessor.class);
+    }
+
+    public static class JettyIncludeProcessor extends IncludeProcessor
+    {
+        @Override
+        public boolean handles(String target)
+        {
+            return "jetty".equals(target);
+        }
+
+        @Override
+        public void process(Document document, PreprocessorReader reader, String target, Map<String, Object> attributes)
+        {
+            try
+            {
+                Path projectPath = Path.of((String)document.getAttribute("projectdir"));
+                Path jettyHome = projectPath.resolve("jetty-home/target/jetty-home").normalize();
+
+                String setupArgs = (String)attributes.get("setupArgs");
+                if (setupArgs != null)
+                    setupArgs += " " + jettyHome.resolve("etc/jetty-halt.xml");
+
+                String args = (String)attributes.get("args");
+                args = args == null ? "" : args + " ";
+                args += jettyHome.resolve("etc/jetty-halt.xml");
+
+                JettyHomeTester jetty = JettyHomeTester.Builder.newInstance()
+                    .jettyHome(jettyHome)
+                    .mavenLocalRepository((String)document.getAttribute("mavenrepository"))
+                    .build();
+
+                // Run first the setup arguments, then the normal arguments.
+                String args0 = setupArgs != null ? setupArgs : args;
+                String args1 = setupArgs != null ? args : null;
+
+                try (JettyHomeTester.Run run0 = jetty.start(args0.split(" ")))
+                {
+                    run0.awaitFor(15, TimeUnit.SECONDS);
+
+                    String output;
+                    if (args1 != null)
+                    {
+                        try (JettyHomeTester.Run run1 = jetty.start(args1.split(" ")))
+                        {
+                            run1.awaitFor(15, TimeUnit.SECONDS);
+                            output = captureOutput(attributes, jetty, run1);
+                        }
+                    }
+                    else
+                    {
+                        output = captureOutput(attributes, jetty, run0);
+                    }
+
+                    reader.push_include(output, "jettyHome_run", target, 1, attributes);
+                }
```

Comment:
I find this code flow very confusing.  Why have arg0 sometimes be args and sometimes be setupArgs?

Why not just if setupArgs are non null, run without capture, then ifargs are non null run with capture.

Reply:
If `setupArgs` are non null, you need to run the `setupArgs` and then, nested, run `args` with capture -- I can't just run without capture.
Furthermore, I would duplicate the try-with-resources block that executes the normal run.
I prefer to fiddle with the arguments, number them, then do the runs.

Reply:
If `setupArgs` are non-null, I can't run without capture: I need to nest a normal run and capture that.
I would duplicate the try-with-resources for the normal run which I don't want to as they are the important bit of logic.
I prefer to fiddle with the arguments and only have 2 try-with-resources.

Reply:
```suggestion
                if (setupArgs != null)
                {
                    try (JettyHomeTester.Run runSetup = jetty.start(setupArgs.split(" ")))
                    {
                        run0.awaitFor(15, TimeUnit.SECONDS);
                    }
                }
                if (args != null)
                {
                    try (JettyHomeTester.Run run = jetty.start(args.split(" ")))
                    {
                        run.awaitFor(15, TimeUnit.SECONDS);
                        String output = captureOutput(attributes, jetty, run);
                        reader.push_include(output, "jettyHome_run", target, 1, attributes);
                    }
                }
```